### PR TITLE
[Keyboard] Add Foundation to QMK

### PR DESCRIPTION
**Pull request by mechlovin**
## Description

Add Foundation PCB to QMK.
F row less layout, APM32F103, rotary encoder.
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/mechlovin/foundation/config.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#ifdef RGB_DI_PIN
+#    define RGBLED_NUM 22
+#    define RGBLIGHT_HUE_STEP 8
+#    define RGBLIGHT_SAT_STEP 8
+#    define RGBLIGHT_VAL_STEP 8
+#    define RGBLIGHT_LIMIT_VAL 255 /* The maximum brightness level */
+#    define RGBLIGHT_SLEEP  /* If defined, the RGB lighting will be switched off when the host goes to sleep */
+/*== all animations enable ==*/
+//#    define RGBLIGHT_ANIMATIONS
+/*== or choose animations ==*/
```
Comment (drashna) [inline keyboards/mechlovin/foundation/config.h:65]:
```suggestion
```

File: keyboards/mechlovin/foundation/foundation.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    encoder_action_unregister();
+    matrix_scan_user();
+}
+
+bool encoder_update_kb(uint8_t index, bool clockwise)
+{
+    encoder_action_register(index, clockwise);
+    return true;
+};
+#else
```
Comment (drashna) [inline keyboards/mechlovin/foundation/foundation.c:61]:
```suggestion
#ifdef ENCODER_ENABLE
```

File: keyboards/mechlovin/foundation/keymaps/via/keymap.c [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    [3] = LAYOUT_all(
+        KC_TRNS,  KC_TRNS,  KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,   KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS,  KC_TRNS,  KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,            KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS,  KC_TRNS,  KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,                     KC_TRNS, KC_TRNS,
+        KC_TRNS,  KC_TRNS,  KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,                     KC_TRNS,
+        KC_TRNS,  KC_TRNS,  KC_TRNS,  KC_TRNS,                            KC_TRNS,                            KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,             KC_TRNS, KC_TRNS, KC_TRNS
+    ),
+
+};
+
```
Comment (drashna) [inline keyboards/mechlovin/foundation/keymaps/via/keymap.c:49]:
```suggestion
#ifdef VIA_ENABLE
static uint8_t encoder_state[ENCODERS] = {0};
static keypos_t encoder_cw[ENCODERS] = ENCODERS_CW_KEY;
static keypos_t encoder_ccw[ENCODERS] = ENCODERS_CCW_KEY;

void encoder_action_unregister(void) {
    for (int index = 0; index < ENCODERS; ++index) {
        if (encoder_state[index]) {
            keyevent_t encoder_event = (keyevent_t){
                .key = encoder_state[index] >> 1 ? encoder_cw[index] : encoder_ccw[index],
                .pressed = false,
                .time = (timer_read() | 1)};
            encoder_state[index] = 0;
            action_exec(encoder_event);
        }
    }
}

void encoder_action_register(uint8_t index, bool clockwise) {
    keyevent_t encoder_event = (keyevent_t){
        .key = clockwise ? encoder_cw[index] : encoder_ccw[index],
        .pressed = true,
        .time = (timer_read() | 1)};
    encoder_state[index] = (clockwise ^ 1) | (clockwise << 1);
    action_exec(encoder_event);
}

void matrix_scan_user(void) {
    encoder_action_unregister();
}

bool encoder_update_user(uint8_t index, bool clockwise)
{
    encoder_action_register(index, clockwise);
    return true;
};
```

File: keyboards/mechlovin/foundation/foundation.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    { K10,   K11,   K12,   K13,   K14,   K15,   K16,   K17,   K18,   K19,   K1A,   K1B,   K1C,   K1D,   K1E,   K1F,     K1G,   K1H }, \
+    { K20,   K21,   K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C,   K2D,   K2E,   K2F,     K2G,   K2H }, \
+    { K30,   K31,   K32,   K33,   K34,   K35,   K36,   K37,   K38,   K39,   K3A,   K3B,   K3C,   K3D,   KC_NO, KC_NO,   K3G, KC_NO }, \
+    { K40,   K41,   K42,   K43, KC_NO, KC_NO, KC_NO,   K47, KC_NO, KC_NO, KC_NO,   K4B,   K4C,   K4D,   K4E,   K4F,     K4G,   K4H }, \
+}
+
+#define LAYOUT_tkl_nofrow_ansi( \
+    K00, K01, K02, K03, K04, K05, K06, K07, K08, K09, K0A, K0B, K0C, K0D, K0E,      K0F, K0G, K0H, \
+    K10, K11, K12, K13, K14, K15, K16, K17, K18, K19, K1A, K1B, K1C, K1D, K1E,      K1F, K1G, K1H, \
+    K20, K21, K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C,      K2E,           K2G, K2H, \
```
Comment (drashna) [inline keyboards/mechlovin/foundation/foundation.h:37]: K2G and K2H are for the encoder, correct?  
If so, they should be removed. 
```suggestion
    K20, K21, K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C,      K2E,                     \
```

File: keyboards/mechlovin/foundation/foundation.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    { K10,   K11,   K12,   K13,   K14,   K15,   K16,   K17,   K18,   K19,   K1A,   K1B,   K1C,   K1D,   K1E,   K1F,     K1G,   K1H }, \
+    { K20,   K21,   K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C, KC_NO,   K2E, KC_NO,     K2G,   K2H }, \
+    { K30,   K31, KC_NO,   K33,   K34,   K35,   K36,   K37,   K38,   K39,   K3A,   K3B,   K3C,   K3D, KC_NO, KC_NO,     K3G, KC_NO }, \
+    { K40,   K41,   K42,   K43, KC_NO, KC_NO, KC_NO,   K47, KC_NO, KC_NO, KC_NO,   K4B,   K4C,   K4D,   K4E,   K4F,     K4G,   K4H }, \
+}
+
+#define LAYOUT_tkl_nofrow_iso( \
+    K00, K01, K02, K03, K04, K05, K06, K07, K08, K09, K0A, K0B, K0C, K0D, K0E,      K0F, K0G, K0H, \
+    K10, K11, K12, K13, K14, K15, K16, K17, K18, K19, K1A, K1B, K1C, K1D, K1E,      K1F, K1G, K1H, \
+    K20, K21, K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C,      K2E,           K2G, K2H, \
```
Comment (drashna) [inline keyboards/mechlovin/foundation/foundation.h:51]:
```suggestion
    K20, K21, K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C,      K2E,                     \
```

File: keyboards/mechlovin/foundation/foundation.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+#define LAYOUT_tkl_nofrow_ansi( \
+    K00, K01, K02, K03, K04, K05, K06, K07, K08, K09, K0A, K0B, K0C, K0D, K0E,      K0F, K0G, K0H, \
+    K10, K11, K12, K13, K14, K15, K16, K17, K18, K19, K1A, K1B, K1C, K1D, K1E,      K1F, K1G, K1H, \
+    K20, K21, K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C,      K2E,           K2G, K2H, \
+    K30, K31,      K33, K34, K35, K36, K37, K38, K39, K3A, K3B, K3C, K3D,                K3G,       \
+    K40, K41, K42, K43,                K47,                K4B, K4C, K4D, K4E,      K4F, K4G, K4H   \
+) { \
+    { K00,   K01,   K02,   K03,   K04,   K05,   K06,   K07,   K08,   K09,   K0A,   K0B,   K0C,   K0D,   K0E,   K0F,     K0G,   K0H }, \
+    { K10,   K11,   K12,   K13,   K14,   K15,   K16,   K17,   K18,   K19,   K1A,   K1B,   K1C,   K1D,   K1E,   K1F,     K1G,   K1H }, \
+    { K20,   K21,   K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C, KC_NO,   K2E, KC_NO,     K2G,   K2H }, \
```
Comment (drashna) [inline keyboards/mechlovin/foundation/foundation.h:43]:
```suggestion
    { K20,   K21,   K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C, KC_NO,   K2E, KC_NO,     KC_NO,   KC_NO }, \
```

File: keyboards/mechlovin/foundation/foundation.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+#define LAYOUT_tkl_nofrow_iso( \
+    K00, K01, K02, K03, K04, K05, K06, K07, K08, K09, K0A, K0B, K0C, K0D, K0E,      K0F, K0G, K0H, \
+    K10, K11, K12, K13, K14, K15, K16, K17, K18, K19, K1A, K1B, K1C, K1D, K1E,      K1F, K1G, K1H, \
+    K20, K21, K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C,      K2E,           K2G, K2H, \
+    K30, K31, K32, K33, K34, K35, K36, K37, K38, K39, K3A, K3B, K3C, K3D,                K3G,       \
+    K40, K41, K42, K43,                K47,                K4B, K4C, K4D, K4E,      K4F, K4G, K4H   \
+) { \
+    { K00,   K01,   K02,   K03,   K04,   K05,   K06,   K07,   K08,   K09,   K0A,   K0B,   K0C,   K0D,   K0E,   K0F,     K0G,   K0H }, \
+    { K10,   K11,   K12,   K13,   K14,   K15,   K16,   K17,   K18,   K19,   K1A,   K1B,   K1C,   K1D,   K1E,   K1F,     K1G,   K1H }, \
+    { K20,   K21,   K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C, KC_NO,   K2E, KC_NO,     K2G,   K2H }, \
```
Comment (drashna) [inline keyboards/mechlovin/foundation/foundation.h:57]:
```suggestion
    { K20,   K21,   K22,   K23,   K24,   K25,   K26,   K27,   K28,   K29,   K2A,   K2B,   K2C, KC_NO,   K2E, KC_NO,     KC_NO, KC_NO }, \
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/mechlovin/foundation/foundation.h (hunk 4 above)
Comment (mechlovin) [reply keyboards/mechlovin/foundation/foundation.h]: Yes, they are for the encoder.
Thanks.

**Comment by drashna** [comment]
Lint: 

> ☒ mechlovin/foundation: Number of keys for LAYOUT_tkl_nofrow_ansi does not match! info.json specifies 78 keys, C macro specifies 76
> ☒ mechlovin/foundation: Number of keys for LAYOUT_tkl_nofrow_iso does not match! info.json specifies 79 keys, C macro specifies 77

**Review by drashna — APPROVED** [review]

**Review by keyboard-magpie — COMMENTED** [review]
Do you want to change this to [encoder map](https://docs.qmk.fm/#/feature_encoders?id=encoder-map) before it's merged @mechlovin

**Comment by mechlovin** [comment]
Yes, please wait me for few hour.

**Review by keyboard-magpie — APPROVED** [review]

**Comment by mechlovin** [comment]
Encoder map has been enabled.
Thanks

**Pull request merged**